arbnode/message_pruner.go

```diff
@@ -0,0 +1,187 @@
+	latestConfirmedNode, err := m.staker.Rollup().LatestConfirmed(
+		&bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
+		})
+	if err != nil {
+		log.Error("error getting latest confirmed node: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	nodeInfo, err := m.staker.Rollup().LookupNode(ctx, latestConfirmedNode)
+	if err != nil {
+		log.Error("error getting latest confirmed node info: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	endBatchCount := nodeInfo.Assertion.AfterState.GlobalState.Batch
+	if endBatchCount == 0 {
+		return m.config().MessagePruneInterval
+	}
+	endBatchMetadata, err := m.inboxTracker.GetBatchMetadata(endBatchCount - 1)
+	if err != nil {
+		log.Error("error getting last batch metadata: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	deleteOldMessageFromDB(endBatchCount, endBatchMetadata, m.inboxTracker.db, m.transactionStreamer.db)
+	return m.config().MessagePruneInterval
+}
+
+func deleteOldMessageFromDB(endBatchCount uint64, endBatchMetadata BatchMetadata, inboxTrackerDb ethdb.Database, transactionStreamerDb ethdb.Database) {
+	startBatchCount := uint64(1)
+	hasKey, err := inboxTrackerDb.Has(lastPrunedSequencerBatchMetaKey)
+	if err != nil {
+		log.Warn("error checking last pruned batch metadata: %w", err)
+	} else if hasKey {
+		data, err := inboxTrackerDb.Get(lastPrunedSequencerBatchMetaKey)
+		if err != nil {
+			log.Warn("error fetching last pruned batch metadata: %w", err)
+		} else {
+			err = rlp.DecodeBytes(data, &startBatchCount)
+			if err != nil {
+				log.Warn("error decoding last pruned batch metadata: %w", err)
+			}
+		}
+	}
+	if endBatchCount > startBatchCount {
+		err := deleteFromRange(inboxTrackerDb, sequencerBatchMetaPrefix, startBatchCount, endBatchCount-1)
+		if err != nil {
+			log.Error("error deleting batch metadata: %w", err)
+			return
+		}
+		endBatchCountValue, err := rlp.EncodeToBytes(endBatchCount - 1)
+		if err != nil {
+			log.Error("error encoding end batch count: %w", err)
+			return
+		}
+		err = inboxTrackerDb.Put(lastPrunedSequencerBatchMetaKey, endBatchCountValue)
+		if err != nil {
+			log.Error("error storing last pruned batch metadata: %w", err)
+			return
+		}
+	}
+
+	startMessageCount := uint64(1)
+	hasKey, err = transactionStreamerDb.Has(lastPrunedMessageKey)
+	if err != nil {
+		log.Warn("error checking last pruned message: %w", err)
+	} else if hasKey {
+		data, err := transactionStreamerDb.Get(lastPrunedMessageKey)
+		if err != nil {
+			log.Warn("error fetching last pruned message: %w", err)
+		} else {
+			err = rlp.DecodeBytes(data, &startMessageCount)
+			if err != nil {
+				log.Warn("error decoding last pruned message: %w", err)
+			}
+		}
+	}
+	if uint64(endBatchMetadata.MessageCount) > startMessageCount {
+		err := deleteFromRange(transactionStreamerDb, messagePrefix, startMessageCount, uint64(endBatchMetadata.MessageCount)-1)
```

Comment:
Let's make an info log whenever we prune anything (ideally we'd collate everything we pruned into a single log line). It should only happen every hour, because new nodes get confirmed every hour, so even though message pruning happens every minute it will only take effect every hour.

Reply:
done.